### PR TITLE
Make ZeroPadding2D optionally asymmetric

### PR DESCRIPTION
**Pull request by lukovkin**
I propose to implement an asymmetric padding as an option in the ZeroPadding2D.

Reasons:
- From my experience it could be definitely useful when using convolutional layers for the time series data. At some degree it's covered in #531.
- Probably it could also be useful for the image processing, but I cannot offer an example.

Open questions:
- May be it's better to add a separate layer for asymmetric zero padding instead of changing the existing ZeroPadding2D?
- There's a possible confusion in the existing ZeroPadding2D (see https://github.com/fchollet/keras/blob/master/keras/layers/convolutional.py#L1345-L1346) - dimensions are called width and height, respectively, but in all convolutional layers respective dimensions are rows, cols (vice versa). Should I leave the as is or change?
- May be it makes sense to do the same thing for ZeroPadding1D simultaneously? The reason why I've started with 2D is that 1D convolution is a kind of not 1st class citizen - e.g. it lacks Deconvolution, AtrousConvolution implementations, etc.

### CODE REVIEW

**Comment by crowsonkb** [comment]
> Probably it could also be useful for the image processing, but I cannot offer an example.

I've wanted it before for making the output of Deconvolution2D layers the right size.

**Comment by lukovkin** [comment]
Travis failed at the test of the model saving in Python 2.7 again, that looks to have nothing in common with the changes made.

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -821,6 +821,20 @@ def spatial_2d_padding(x, padding=(1, 1), dim_ordering='th'):
     return tf.pad(x, pattern)
 
 
+def asymmetric_spatial_2d_padding(x, padding=(1, 1, 1, 1), dim_ordering='th'):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Can we have a better API than `padding=(1, 1, 1, 1)`? Maybe `padding=((1, 1), (1, 1))` (but that one is not very good easier). Please think of something. Maybe multiple arguments.

File: keras/layers/convolutional.py [hunk 2]
```diff
@@ -1331,26 +1331,30 @@ class ZeroPadding2D(Layer):
         (samples, depth, first_padded_axis, second_padded_axis)
     '''
 
-    def __init__(self, padding=(1, 1), dim_ordering='default', **kwargs):
+    def __init__(self, padding=(1, 1, 1, 1), dim_ordering='default', **kwargs):
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Same.

**Comment by lukovkin** [comment]
Faced some issues after the default image dimensions changed to the 'tf', will try to resolve.

**Comment by lukovkin** [comment]
OK, looks like I've fixed a bug, remaining errors are in the other tests.

So, paddings for the asymmetric padding are to be passed as a separate arguments (top_pad, bottom_pad, left_pad, right_pad).
Option to pass the symmetric 'padding' as a tuple of 2 is kept for the compatibility. If 'padding' has value different from (1, 1) it overrides other arguments.

If it's OK in general, I'll do some cleanup and that's it.

**Comment by lukovkin** [comment]
Ah, wait, I'll do the asymmetric ZeroPadding1D too, as soon as now we have a dilated convolution 1D, thanks to #3763

**Comment by lukovkin** [comment]
@fchollet Would you be so kind to provide a feedback?

**Comment by fchollet** [comment]
@lukovkin it looks like this PR includes a lot of changes that shouldn't be included. Sync to master and update the PR.

**Comment by lukovkin** [comment]
@fchollet OK

**Comment by lukovkin** [comment]
@fchollet Done syncing and cleanup.

**Comment by lukovkin** [comment]
@fchollet Would you be so kind to provide a feedback?

**Comment by Itaydal** [comment]
Hi @fchollet, just came up with the same need for asymmetric padding (for time series analysis as well)
Could you please take a look into it?

P.S
I know that i'm stating the obvious, the Keras is awesome, thanks for it!

File: keras/layers/convolutional.py [hunk 3]
```diff
@@ -1459,46 +1484,79 @@ class ZeroPadding2D(Layer):
...
-    def __init__(self, padding=(1, 1), dim_ordering='default', **kwargs):
+    def __init__(self,
+                 padding=(1, 1),
+                 top_pad=1,
+                 bottom_pad=1,
+                 left_pad=1,
+                 right_pad=1,
+                 dim_ordering='default',
+                 **kwargs):
+        import sys
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Imports should go at the top of the file. Also this import is not used anywhere.

File: keras/backend/theano_backend.py [hunk 4]
```diff
@@ -604,6 +619,41 @@ def spatial_2d_padding(x, padding=(1, 1), dim_ordering=_IMAGE_DIM_ORDERING):
     return T.set_subtensor(output[indices], x)
 
 
+def asymmetric_spatial_2d_padding(x, top_pad=1, bottom_pad=1, left_pad=1, right_pad=1, dim_ordering=_IMAGE_DIM_ORDERING):
+    '''Pad the rows and columns of a 4D tensor
+    with "top_pad", "bottom_pad", "left_pad", "right_pad"  (resp.) zeros rows on top, bottom; cols on left, right.
+    '''
+    from theano import tensor as T
+    import sys
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: First import is already present in the file, second import is not necessary.

File: keras/layers/convolutional.py [hunk 5]
```diff
@@ -1449,7 +1464,17 @@ class ZeroPadding2D(Layer):
     # Arguments
         padding: tuple of int (length 2)
             How many zeros to add at the beginning and end of
-            the 2 padding dimensions (axis 3 and 4).
+            the 2 padding dimensions (rows and cols) symmetrically.
+            Setting padding to the values different from (1, 1)
+            overrides top_pad, bottom_pad; and left_pad, right_pad arguments.
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Really not a fan of this API. Maybe we should go with a single polymorphic argument, `padding`. If tuple of length 2, it's symmetric padding. If length 4, it's in order `top_pad`, `bottom_pad`, `left_pad`, `right_pad `. It can also be dictionary with keys `top_pad`, `bottom_pad`, and `left_pad`, `right_pad`; any missing keys defaults to 0. Any unexpected key or tuple of unexpected size triggers a helpful exception.

What do you think?

File: tests/keras/layers/test_convolutional.py [hunk 6]
```diff
@@ -369,8 +369,14 @@ def test_zero_padding_2d():
     stack_size = 2
     input_nb_row = 11
     input_nb_col = 12
+    dim_ordering = K.image_dim_ordering()
+    print("Dim ordering: {}".format(dim_ordering))
```
Comment (fchollet) [inline tests/keras/layers/test_convolutional.py]: Inconsistent use of `"` (should be `'`). Also not clear that this statement is necessary.

File: keras/layers/convolutional.py [hunk 7]
```diff
@@ -1423,22 +1431,29 @@ class ZeroPadding1D(Layer):
         3D tensor with shape (samples, padded_axis, features)
     '''
 
-    def __init__(self, padding=1, **kwargs):
+    def __init__(self, padding=1, left_pad=1, right_pad=1, **kwargs):
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Same here, I think we should have a polymorphic padding argument that could be `int`, tuple of size 2, or dictionary.

**Comment by fchollet** [comment]
Sorry for the delay in reviewing.

**Comment by Itaydal** [comment]
Hi, @fchollet @lukovkin 
I've made the requested changes, and i want to push for reviewing, but i get push error with access denied message at my GitKraken.
Could you guys please guide me of what should I do? I'm new to the open source community :)

Thanks!

File: keras/layers/convolutional.py (hunk 5 above)
Comment (lukovkin) [inline keras/layers/convolutional.py]: @fchollet It was a polymorphic initially, but you asked me to come up with another solution ;)
No problem, I think I'll go for a tuple of length 2 and a dictionary.

Comment (lukovkin) [inline keras/layers/convolutional.py]: @fchollet One question, if I'll leave an option to pass the padding as a tuple of length 2, it should also be left as a default argument of value (1, 1). But in this case it will contradict with the defaults for the dictionary (missing -> zero). I would set defaults for the dict to 1's too, I can change later if you wouldn't agree.

Comment (Itaydal) [inline keras/layers/convolutional.py]: would you like to take a look at what i've done? (attached)
also, i think that you have had a bug on the ordering of the 2D padding class at get_output_shape_for, you mixed up hight/width from rows/cols.

[convolutional.py.zip](https://github.com/fchollet/keras/files/522910/convolutional.py.zip)

Comment (lukovkin) [inline keras/layers/convolutional.py]: @Itaydal Thank you, I'll take a look.

Comment (lukovkin) [inline keras/layers/convolutional.py]: @Itaydal I've took a look at your code and changed some docstrings accordingly.
Regarding ordering, can you elaborate?
I use (samples, channels, rows, cols) ordering for 'th' and (samples, rows, cols, channels) for 'tf'.
I use top_pad and bottom_pad for rows dimension and left_pad and right_pad for cols dimension.

File: keras/layers/convolutional.py (hunk 7 above)
Comment (lukovkin) [inline keras/layers/convolutional.py]: @fchollet Done

File: keras/layers/convolutional.py (hunk 5 above)
Comment (lukovkin) [inline keras/layers/convolutional.py]: @fchollet Done.

File: keras/layers/convolutional.py (hunk 3 above)
Comment (lukovkin) [inline keras/layers/convolutional.py]: @fchollet Removed it.

File: keras/layers/convolutional.py [hunk 8]
```diff
@@ -1426,16 +1432,27 @@ class ZeroPadding1D(Layer):
     def __init__(self, padding=1, **kwargs):
         super(ZeroPadding1D, self).__init__(**kwargs)
         self.padding = padding
+
+        if isinstance(padding, int):
+            self.left_pad = padding
+            self.right_pad = padding
+        elif isinstance(padding, dict):
+            self.left_pad = padding.get('left_pad', 1)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Default values should be 0 (i.e. if argument isn't specified then no padding occurs on that side).

File: keras/layers/convolutional.py [hunk 9]
```diff
@@ -1426,16 +1432,27 @@ class ZeroPadding1D(Layer):
...
+            self.right_pad = padding.get('right_pad', 1)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Same.

File: keras/layers/convolutional.py [hunk 10]
```diff
@@ -1412,9 +1412,15 @@ class ZeroPadding1D(Layer):
...
-        padding: int
+        padding: int or tuple of int (length 2) or dictionary
+            For symmetric padding: int
             How many zeros to add at the beginning and end of
             the padding dimension (axis 1).
+            For asymmetric padding: tuple of int (length 2)
+            How many zeros to add at the beginning and at the end of
+            the padding dimension (left_pad, right_pad) or
+            {'left_pad': left_pad, 'right_pad': right_pad}.
+            If any key is missing, default value of 1 will be used for the missing key.
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Change to 0. Also please format code elements in docstring with "`" (tuples. dicts).

File: keras/layers/convolutional.py [hunk 11]
```diff
@@ -1447,9 +1464,16 @@ class ZeroPadding2D(Layer):
...
+            For symmetric padding tuple of int (length 2)
             How many zeros to add at the beginning and end of
-            the 2 padding dimensions (axis 3 and 4).
+            the 2 padding dimensions (rows and cols).
+            For asymmetric padding tuple of int (length 4)
+            How many zeros to add at the beginning and at the end of
+            the 2 padding dimensions (rows and cols).
+            (top_pad, bottom_pad, left_pad, right_pad) or
+            {'top_pad': top_pad, 'bottom_pad': bottom_pad, 'left_pad': left_pad, 'right_pad': right_pad}
+            If any key is missing, default value of 1 will be used for the missing key.
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Change to 0. Also please format code elements in docstring with "`" (tuples. dicts).

File: keras/layers/convolutional.py [hunk 12]
```diff
@@ -1459,43 +1483,75 @@ class ZeroPadding2D(Layer):
...
         if dim_ordering == 'default':
             dim_ordering = K.image_dim_ordering()
-        self.padding = tuple(padding)
+
+        self.padding = padding
+        try:
+            self.top_pad = padding.get('top_pad', 1)
+            self.bottom_pad = padding.get('bottom_pad', 1)
+            self.left_pad = padding.get('left_pad', 1)
+            self.right_pad = padding.get('right_pad', 1)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Change to 0.

Comment (fchollet) [inline keras/layers/convolutional.py]: Additionally, raise a `ValueError` with a helpful message in case an unexpected key is found in the dict. This is important because people may have typos in their dict argument and they would never notice.

File: keras/layers/convolutional.py (hunk 8 above)
Comment (fchollet) [inline keras/layers/convolutional.py]: Additionally, raise a `ValueError` with a helpful message in case an unexpected key is found in the dict. This is important because people may have typos in their dict argument and they would never notice.

File: keras/layers/convolutional.py (hunk 9 above)
Comment (lukovkin) [inline keras/layers/convolutional.py]: Done.

File: keras/layers/convolutional.py (hunk 12 above)
Comment (lukovkin) [inline keras/layers/convolutional.py]: Done.

**Comment by lukovkin** [comment]
Also added test for the ZeroPadding1D to the test_convolutional.py.

**Comment by fchollet** [comment]
Thanks, LGTM
